### PR TITLE
Keeping Grappa up-to-date

### PR DESCRIPTION
**Pull request by kelemenb-ewit**
I have made some modifications on Grappa to be able to use it with Graphviz version 2.40.1 (20161225.0304) graph filter command line tools.
While I was here I have made some modification in the error displaying as well.
Please review my changes, and consider merging them to the "original" branch to create a new, up-to-date grappa.jar release!

### CODE REVIEW

File: src/main/java/att/grappa/Lexer.java [hunk 1]
```diff
@@ -522,6 +552,11 @@ private Symbol real_next_token() throws IOException
                 return do_quote_string();
             }
 
+            // equals may not followed by a quoted text -- let's read this word
```
Comment (sdumitriu) [inline src/main/java/att/grappa/Lexer.java]: This comment isn't clear enough, can you rephrase it?

File: src/main/java/att/grappa/AwtExceptionDisplay.java [hunk 2]
```diff
@@ -0,0 +1,203 @@
...
+    public AwtExceptionDisplay(String title)
+    {
+        this.title = title;
+    }
+
+    /**
+     * Pops up the frame and displays information on the supplied exception. Initially, a text area displays the message
+     * associated with the exception. By pressing a button, an end-user can view a stack trace as well.
+     *
+     * @param ex the exception about which informtaion is to be displayed.
```
Comment (sdumitriu) [inline src/main/java/att/grappa/AwtExceptionDisplay.java]: You didn't fix the `informtaion` typo here, but you did in the original fine. Can you fix it here too?

File: src/main/java/att/grappa/ExceptionDisplay.java [hunk 3]
```diff
@@ -1,202 +1,23 @@
-/*
```
Comment (sdumitriu) [inline src/main/java/att/grappa/ExceptionDisplay.java:1]: Why did you remove the license header? Can you put it back please?

File: src/main/java/att/grappa/Grappa.java [hunk 4]
```diff
@@ -110,9 +110,15 @@
     /*
      * A instance of ExceptionDisplay used with the (public) convenience methods that follow.
      */
-    private static final ExceptionDisplay exceptionDisplay = new ExceptionDisplay(Grappa.PKG_UPLOW
+    private static ExceptionDisplay exceptionDisplay = new AwtExceptionDisplay(Grappa.PKG_UPLOW
```
Comment (sdumitriu) [inline src/main/java/att/grappa/Grappa.java:113]: Do you have a non-AWT implementation that you wouldn't mind sharing?

File: src/main/java/att/grappa/Element.java [hunk 5]
```diff
@@ -994,8 +994,6 @@ public static String canonString(String input)
         StringBuilder strbuf = new StringBuilder(len + 8);
         char[] array = input.toCharArray();
         boolean has_special = false;
-        boolean isHTML = false;
```
Comment (sdumitriu) [inline src/main/java/att/grappa/Element.java]: HTML labels are supported, according to [the documentation](http://www.graphviz.org/content/node-shapes#html) and a quick test that I ran. I'd say this commit should be skipped. Is there anything breaking on your end?

**Review by sdumitriu — CHANGES_REQUESTED** [review]

File: src/main/java/att/grappa/AwtExceptionDisplay.java (hunk 2 above)
Comment (kelemenb-ewit) [reply src/main/java/att/grappa/AwtExceptionDisplay.java]: I wanted to make the least modification possible, but of course I can do this.

File: src/main/java/att/grappa/Element.java (hunk 5 above)
Comment (kelemenb-ewit) [reply src/main/java/att/grappa/Element.java]: You are probably right. What I'm facing with is, that in our project we put a HTML-like attribute value into the "graph" section (Graph.setAttribute() ), and dot claims for an illegal character here.
It there is no problem, where the HTML-like value is quoted.
But there are some things not clear for me:
 1. Why is is required in the original code to disable quoting for HTML-like attributes? Does it matters something in the graphviz side?
 2. Why do we put HTML-like attributes in the "graph" section.

File: src/main/java/att/grappa/ExceptionDisplay.java (hunk 3 above)
Comment (kelemenb-ewit) [reply src/main/java/att/grappa/ExceptionDisplay.java:1]: Actually the ExceptionDisplay.java is renamed to AwtExceptionDisplay.java. So this ExceptionDisplay.java is a new file, where the author is me.

File: src/main/java/att/grappa/Grappa.java (hunk 4 above)
Comment (kelemenb-ewit) [reply src/main/java/att/grappa/Grappa.java:113]: We do logging with log4j. Including my Log4j exception displaying implementation in the grappa code, would introduce new jar dependencies. I do not recommend that. Probably as an example in some comment section?

File: src/main/java/att/grappa/Element.java (hunk 5 above)
Comment (sdumitriu) [reply src/main/java/att/grappa/Element.java]: HTML attributes are surrounded by `<...>`, for example:

    a [label=<<I>hello</I>>]
    b [label=<
      <B>Longer text</B>
      <BR/>
      <I>Spanning multiple lines</I>
    >]

Comment (sdumitriu) [reply src/main/java/att/grappa/Element.java]: So, in this case `<` and `>` act like the quotes. Using regular quotes works too, but in that case it loses it's special "this is an HTML label" meaning, and using nothing to quote is invalid.

File: src/main/java/att/grappa/Grappa.java (hunk 4 above)
Comment (sdumitriu) [reply src/main/java/att/grappa/Grappa.java:113]: You are right, new dependencies should be avoided at this time.

File: src/main/java/att/grappa/Element.java (hunk 5 above)
Comment (kelemenb-ewit) [reply src/main/java/att/grappa/Element.java]: It is interesting, as we are using a similar form (for an unknown reason), and dot claims error.
Example:
```
digraph "our content" {
  graph [
    layoutAlgorithmConfig = <layout-algorithm-config><edge-mapping /></layout-algorithm-config>,
...
```
dot error:
```
Error: temp/dot.in: syntax error in line 3 near '>'
```
Maybe it accepts forms like this for a label, but for any other attributes?
If you can recommend me a proper solution for that, I will implement it. Or we just can leave this commit out, as we can solve this issue internally.

Comment (kelemenb-ewit) [reply src/main/java/att/grappa/Element.java]: Okay, now I understand. Quoting with < > works as you have described. Thx! I have reverted this commit.

**Review by sdumitriu — APPROVED** [review]

File: src/main/java/att/grappa/ExceptionDisplay.java (hunk 3 above)
Comment (sdumitriu) [reply src/main/java/att/grappa/ExceptionDisplay.java:1]: This still needs a license header, otherwise it might be assumed that the file is uncopyrighted, i.e. "public domain". I merged this PR, so I assumed that you're OK to include this under the same license as every other file in the project.

File: src/main/java/att/grappa/Lexer.java [hunk 6]
```diff
@@ -389,6 +389,37 @@ private Symbol do_quote_string() throws IOException
...
+     * and dots.
+     * The routine returns a Symbol object suitable for return by the scanner.
+     */
+    private Symbol do_unquote_string() throws IOException
+    {
+        String result_str;
+
+        synchronized (this.cmnstrbuf) {
+            this.cmnstrbuf.delete(0, this.cmnstrbuf.length()); // faster than cmnstrbuf.setLength(0)!
+            // save chars until we see a double quote
```
Comment (sdumitriu) [inline src/main/java/att/grappa/Lexer.java:403]: Not correct in this case, bad copy/paste.

File: src/main/java/att/grappa/Lexer.java [hunk 7]
```diff
@@ -522,6 +553,14 @@ private Symbol real_next_token() throws IOException
                 return do_quote_string();
             }
 
+            // Auoted values are processed with the section above.
```
Comment (sdumitriu) [inline src/main/java/att/grappa/Lexer.java:556]: Typo: `Quoted`
